crates/core/src/network/lazy_connection.rs

```diff
@@ -0,0 +1,460 @@
+                    sort_code,
+                    payload,
+                    version: protocol_version,
+                    notifier,
+                    #[cfg(feature = "test-util")]
+                    header: None,
+                }
+            }
+        }
+    }
+}
+
+struct ConnectionTask<T> {
+    auto_reconnect: bool,
+    peer: GenerationalNodeId,
+    swimlane: Swimlane,
+    is_connected: Arc<AtomicBool>,
+    network_sender: T,
+    msg_rx: mpsc::Receiver<OutgoingOp>,
+}
+
+impl<T: NetworkSender> ConnectionTask<T> {
+    async fn run(self) {
+        let Self {
+            auto_reconnect,
+            peer,
+            swimlane,
+            is_connected,
+            network_sender,
+            mut msg_rx,
+        } = self;
+        let mut inflight_msg: Option<OutgoingOp> = None;
+        'connect: loop {
+            if msg_rx.is_closed() {
+                return;
+            }
+            is_connected.store(false, Ordering::Relaxed);
+            // If this fails with unrecoverable error, we don't try and reconnect, instead we update the
+            // state and return.
+            let connection = match network_sender.get_connection(peer, swimlane).await {
+                Ok(connection) => connection,
+                // We fail immediately if we are not allowed to reconnect anyway.
+                Err(err) if !auto_reconnect => {
+                    debug!(%swimlane, %peer, "Connection failed and we were not asked to reconnect: {}", err);
+                    // don't bother if we were asked to not retry.
+                    return;
+                }
+                // on failure, we should see throttled in the subsequent attempt
+                Err(err @ ConnectError::Transport(_)) => {
+                    trace!(%swimlane, %peer, "Cannot connect, will retry: {}", err);
+                    continue 'connect;
+                }
+                // on failure, we should see throttled in the subsequent attempt
+                Err(err @ ConnectError::Throttled(dur)) => {
+                    trace!(%swimlane, %peer, "Cannot connect, will retry later: {}", err);
+                    // Should we wait the whole duration? probably not, as we might be able to
+                    // connect earlier. This can happen if that peer connected to us with a
+                    // bidirectional connection, or if failure detector decided that this node is
+                    // now reachable. Therefore, we divide the duration by 2.
+                    tokio::time::sleep(dur / 2).await;
+                    continue 'connect;
+                }
+                Err(
+                    // We don't know this node-id yet, but we might in the future, keep retrying.
+                    err @ ConnectError::Discovery(DiscoveryError::UnknownNodeId(_)),
+                ) => {
+                    trace!(%swimlane, %peer, "Cannot to connect, will retry: {}", err);
+                    continue 'connect;
```

Comment:
Could this be a busy loop?

Reply:
Yes, good catch. I'll add a hardcoded 500ms sleep here.